eth/downloader/skeleton.go

```diff
@@ -722,11 +722,22 @@ func (s *skeleton) executeTask(peer *peerConnection, req *headerRequest) {
 
 	case <-timeoutTimer.C:
 		// Header retrieval timed out, update the metrics
-		peer.log.Trace("Header request timed out", "elapsed", ttl)
+		peer.log.Warn("Header request timed out, dropping peer", "elapsed", ttl)
 		headerTimeoutMeter.Mark(1)
 		s.peers.rates.Update(peer.id, eth.BlockHeadersMsg, 0, 0)
 		s.scheduleRevertRequest(req)
 
+		// At this point we either need to drop the offending peer, or we need a
+		// mechanism to allow waiting for the response and not cancel it. For now
+		// lets go with dropping since the header sizes are deterministic and the
+		// beacon sync runs exclusive (downloader is idle) so there should be no
+		// other load to make timeouts probable. If we notice that timeouts happen
+		// more often than we'd like, we can introduce a tracker for the requests
+		// gone stale and monitor them. However, in that case too, we need a way
+		// to protect against malicious peers never responding, so it would need
+		// a second, hard-timeout mechanism.
+		s.drop(peer.id)
```

Comment:
As it's implemented here, the timeout is based on `ttl := s.peers.rates.TargetTimeout()`. So this might feasibly happen: 

- One super-fast peer with good connection obtains a ttl of `5ms`. 
- 10 laggy peers obtain ttls of `500ms`. 
- The fast peer hits a slow-down, and takes `100ms` for a delivery. 
- We drop it the fast peer, and only keep the slow ones. 

Haven't checked if the numbers I made up make any sense, but you probably understand the idea.

Reply:
The message rate tracker tries to track the *median* RTT across all peers. The timeout is the median RTT multiplied by some confidence number.

In your specific scenario, if you have one super fast peer and many laggy ones, the TTL will be the laggy value since that's the median. Even if your fast peer returns a bit slow once, it should still be fine.